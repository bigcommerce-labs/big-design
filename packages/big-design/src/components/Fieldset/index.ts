export { FieldsetDescription } from './Description';
export { Fieldset, type FieldsetProps } from './Fieldset';
export { FieldsetLegend } from './Legend';
