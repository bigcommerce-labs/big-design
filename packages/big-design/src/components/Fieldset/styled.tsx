import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

export const StyledFieldset = styled.fieldset`
  border: none;
  margin: 0 0 ${({ theme }) => theme.spacing.xLarge};
  padding: 0;

  &:last-child {
    margin: 0;
  }
`;

StyledFieldset.defaultProps = { theme: defaultTheme };
