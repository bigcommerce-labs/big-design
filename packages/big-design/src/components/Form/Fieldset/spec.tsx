import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';

import { Fieldset } from './index';

test('renders a fieldset tag', () => {
  const { container } = render(<Fieldset />);

  expect(container.querySelector('fieldset')).toBeInTheDocument();
});

test('renders legend', () => {
  const legendText = 'legend text';
  const { container } = render(<Fieldset legend={legendText} />);
  const legend = container.querySelector('legend') as HTMLLegendElement;

  expect(legend).toBeInTheDocument();
});

test('renders description', () => {
  const descriptionText = 'description text';
  const { queryByText } = render(<Fieldset description={descriptionText} />);

  expect(queryByText(descriptionText)).toBeInTheDocument();
});

test('accepts a Legend Component', () => {
  const CustomLegend = (
    <Fieldset.Legend>
      This is a custom legend
      <a href="#" data-testid="test">
        has a url
      </a>
    </Fieldset.Legend>
  );

  const { queryByTestId } = render(<Fieldset legend={CustomLegend} />);

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Legend Components', () => {
  const NotALegend = (
    <div>
      This is not a custom legend
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(<Fieldset legend={NotALegend} />);

  expect(queryByTestId('test')).not.toBeInTheDocument();
});

test('accepts a Description Component', () => {
  const CustomDescription = (
    <Fieldset.Description>
      This is a custom Description
      <a href="#" data-testid="test">
        has a url
      </a>
    </Fieldset.Description>
  );

  const { queryByTestId } = render(<Fieldset description={CustomDescription} />);

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Description Components', () => {
  const NotADescription = (
    <div>
      This is a not custom description
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(<Fieldset description={NotADescription} />);

  expect(queryByTestId('test')).not.toBeInTheDocument();
});
