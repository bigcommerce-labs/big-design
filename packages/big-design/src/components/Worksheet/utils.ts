import { Cell, WorksheetError, WorksheetItem } from './types';

export const mergeCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  newCells.reduce(
    (accum, newCell) => {
      const index = oldCells.findIndex(
        (oldCell) => oldCell.rowIndex === newCell.rowIndex && oldCell.columnIndex === newCell.columnIndex,
      );

      if (index > -1) {
        accum[index] = newCell;

        return accum;
      }

      return accum.concat(newCell);
    },
    [...oldCells], // Note: returns a new array every time
  );

export const deleteCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  oldCells.filter(
    (oldCell) =>
      !newCells.find((newCell) => newCell.columnIndex === oldCell.columnIndex && newCell.rowIndex === oldCell.rowIndex),
  );

export const editedRows = <T extends WorksheetItem>(editedCells: Cell<T>[], rows: T[]) =>
  editedCells.reduce<T[]>((accum, { rowIndex }) => {
    const row = rows[rowIndex];

    // Check to see if the row already exists in accum
    if (accum.find((editedRow) => editedRow === row)) {
      return accum;
    } else {
      // Only append new rows
      return [...accum, row];
    }
  }, []);

export const invalidRows = <T extends WorksheetItem>(invalidCells: Cell<T>[], rows: T[]) => {
  const mapObj = new Map();

  // Create Map with each row and append errors per row
  invalidCells.forEach(({ rowIndex, hash }) => {
    const row = rows[rowIndex];
    if (mapObj.has(row)) {
      const errors = mapObj.get(row);
      mapObj.set(row, new Set([...errors, hash]));
    } else {
      mapObj.set(row, new Set([hash]));
    }
  });

  return Array.from(mapObj).map<WorksheetError<T>>(([item, errors]) => ({
    item,
    errors: Array.from(errors),
  }));
};
