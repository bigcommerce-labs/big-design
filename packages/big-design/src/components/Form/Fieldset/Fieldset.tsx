import React, { memo } from 'react';

import { StyledFieldset, StyledFieldsetDescription, StyledFieldsetLegend } from './styled';

export interface FieldsetProps extends React.FieldsetHTMLAttributes<HTMLFieldSetElement> {
  legend?: React.ReactChild;
  description?: React.ReactChild;
}

export class Fieldset extends React.PureComponent<FieldsetProps> {
  static Legend = memo(StyledFieldsetLegend);
  static Description = memo(StyledFieldsetDescription);

  render() {
    const { className, legend, description, children, style, ...props } = this.props;

    return (
      <StyledFieldset {...props}>
        {this.renderLegend()}
        {this.renderDescription()}
        {children}
      </StyledFieldset>
    );
  }

  private renderLegend() {
    const { legend } = this.props;

    if (typeof legend === 'string') {
      return <Fieldset.Legend>{legend}</Fieldset.Legend>;
    }

    if (React.isValidElement(legend) && legend.type === Fieldset.Legend) {
      return legend;
    }

    return null;
  }

  private renderDescription() {
    const { description } = this.props;

    if (typeof description === 'string') {
      return <Fieldset.Description>{description}</Fieldset.Description>;
    }

    if (React.isValidElement(description) && description.type === Fieldset.Description) {
      return description;
    }

    return null;
  }
}
