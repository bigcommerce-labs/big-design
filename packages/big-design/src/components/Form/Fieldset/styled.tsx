import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled, { DefaultTheme, StyledComponent } from 'styled-components';

import { StyleableH3, StyleableSmall } from '../../Typography/private';

export const StyledFieldset = styled.fieldset`
  border: none;
  margin: 0 0 ${({ theme }) => theme.spacing.xLarge};
  padding: 0;

  &:last-child {
    margin: 0;
  }
`;

export const StyledFieldsetDescription = styled(StyleableSmall)``;

export const StyledFieldsetLegend = styled(StyleableH3).attrs({ as: 'legend' })`
  &:not(:last-child) {
    margin-bottom: ${({ theme }) => theme.spacing.xxSmall};
  }
` as StyledComponent<'legend', DefaultTheme>;

StyledFieldset.defaultProps = { theme: defaultTheme };
StyledFieldsetLegend.defaultProps = { theme: defaultTheme };
StyledFieldsetDescription.defaultProps = { theme: defaultTheme };
