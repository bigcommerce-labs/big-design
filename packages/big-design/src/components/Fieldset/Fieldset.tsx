import React, { memo, useMemo } from 'react';

import { warning } from '../../utils/warning';

import { StyledFieldset } from './styled';
import { FieldsetDescription } from './Description';
import { FieldsetLegend } from './Legend';

export interface FieldsetProps extends React.FieldsetHTMLAttributes<HTMLFieldSetElement> {
  legend?: React.ReactChild;
  description?: React.ReactChild;
}

export const Fieldset: React.FC<FieldsetProps> = memo(
  ({ className, legend, description, children, style, ...props }) => {
    const renderedLegend = useMemo(() => {
      if (typeof legend === 'string') {
        return <FieldsetLegend>{legend}</FieldsetLegend>;
      }

      if (React.isValidElement(legend) && legend.type === FieldsetLegend) {
        return legend;
      }

      if (!legend) {
        return null;
      }

      warning('legend must be either a string or a FieldsetLegend component.');
    }, [legend]);

    const renderedDescription = useMemo(() => {
      if (typeof description === 'string') {
        return <FieldsetDescription>{description}</FieldsetDescription>;
      }

      if (React.isValidElement(description) && description.type === FieldsetDescription) {
        return description;
      }

      if (!description) {
        return null;
      }

      warning('description must be either a string or a FieldsetDescription component.');
    }, [description]);

    return (
      <StyledFieldset {...props}>
        {renderedLegend}
        {renderedDescription}
        {children}
      </StyledFieldset>
    );
  },
);
