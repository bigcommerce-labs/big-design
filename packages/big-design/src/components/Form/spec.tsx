import { fireEvent, render } from '@test/utils';
import 'jest-styled-components';
import React from 'react';

import { Fieldset } from '../Fieldset';
import { Input } from '../Input';

import { Form, FormGroup } from './index';

test('forwards ref', () => {
  const ref = React.createRef<HTMLFormElement>();

  const { container } = render(<Form ref={ref} />);
  const form = container.querySelector('form');

  expect(form).toBe(ref.current);
});

test('calls onSubmit', () => {
  const onSubmit = jest.fn();
  const { container } = render(<Form onSubmit={onSubmit} />);
  const form = container.querySelector('form') as HTMLFormElement;

  fireEvent.submit(form);

  expect(onSubmit).toHaveBeenCalled();
});

test('simple form render', () => {
  const { container } = render(
    <Form>
      <Fieldset
        legend="Primary contact"
        description="Minim velit quis aute adipisicing adipisicing do do exercitation cupidatat enim ex voluptate consequat labore."
      >
        <FormGroup>
          <Input
            label="First Name"
            description="This is an example description for First Name"
            placeholder="Placeholder text"
          />
        </FormGroup>

        <FormGroup>
          <Input
            label="Middle Name"
            description="This is an example description for Last Name. Featuring a Left Icon."
            placeholder="Placeholder text"
          />
        </FormGroup>
      </Fieldset>
    </Form>,
  );

  expect(container.firstChild).toMatchSnapshot();
});
