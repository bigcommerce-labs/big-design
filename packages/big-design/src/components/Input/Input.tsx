import hoistNonReactStatics from 'hoist-non-react-statics';
import React, { Ref } from 'react';

import { uniqueId } from '../../utils';
import { Chip } from '../Chip';
import { Form } from '../Form';
import { Small } from '../Typography';

import { StyledIconWrapper, StyledInput, StyledInputContent, StyledInputWrapper } from './styled';

interface Props {
  chips?: string[];
  description?: React.ReactChild;
  error?: React.ReactChild | React.ReactChild[];
  iconLeft?: React.ReactChild;
  iconRight?: React.ReactChild;
  label?: React.ReactChild;
  labelId?: string;
  onChipDelete?(chip: string): () => void;
}

interface PrivateProps {
  forwardedRef: Ref<HTMLInputElement>;
}

interface InputState {
  focus: boolean;
}

export type InputProps = Props & React.InputHTMLAttributes<HTMLInputElement>;

class StyleableInput extends React.PureComponent<InputProps & PrivateProps, InputState> {
  static Description = Small;
  static Error = Form.Error;
  static Label = Form.Label;

  readonly state: InputState = {
    focus: false,
  };

  private readonly uniqueId = uniqueId('input_');

  render() {
    const { chips, description, disabled, error, label, labelId, forwardedRef, onChipDelete, ...props } = this.props;
    const id = this.getId();

    return (
      <div>
        {this.renderLabel()}
        {this.renderDescription()}
        <StyledInputWrapper disabled={disabled} error={error} focus={this.state.focus}>
          {this.renderIconLeft()}
          <StyledInputContent chips={chips}>
            {this.renderChips()}
            <StyledInput
              {...props}
              disabled={disabled}
              chips={chips}
              error={error}
              id={id}
              onBlur={this.onInputBlur}
              onFocus={this.onInputFocus}
              ref={forwardedRef}
            />
          </StyledInputContent>

          {this.renderIconRight()}
        </StyledInputWrapper>
      </div>
    );
  }

  private getId() {
    const { id } = this.props;

    return id ? id : this.uniqueId;
  }

  private onInputFocus = (event: React.FocusEvent<HTMLInputElement>) => {
    const { onFocus } = this.props;

    this.setFocus(true);

    return onFocus && onFocus(event);
  };

  private onInputBlur = (event: React.FocusEvent<HTMLInputElement>) => {
    const { onBlur } = this.props;

    this.setFocus(false);

    return onBlur && onBlur(event);
  };

  private renderDescription() {
    const { description } = this.props;

    if (typeof description === 'string') {
      return <Input.Description>{description}</Input.Description>;
    }

    if (React.isValidElement(description) && description.type === Input.Description) {
      return description;
    }

    return null;
  }

  private renderLabel() {
    const { label, labelId, required } = this.props;
    const id = this.getId();

    if (typeof label === 'string') {
      return (
        <Input.Label id={labelId} htmlFor={id} renderOptional={!required}>
          {label}
        </Input.Label>
      );
    }

    if (React.isValidElement(label) && label.type === Input.Label) {
      return React.cloneElement(label as React.ReactElement<React.LabelHTMLAttributes<HTMLLabelElement>>, {
        id: labelId,
        htmlFor: id,
      });
    }

    return null;
  }

  private renderIconLeft() {
    if (!this.props.iconLeft) {
      return null;
    }

    return <StyledIconWrapper>{this.props.iconLeft}</StyledIconWrapper>;
  }

  private renderIconRight() {
    if (!this.props.iconRight) {
      return null;
    }

    return <StyledIconWrapper>{this.props.iconRight}</StyledIconWrapper>;
  }

  private setFocus(toggle: boolean) {
    this.setState({ focus: toggle });
  }

  private renderChips() {
    const { chips, onChipDelete } = this.props;

    if (!chips) {
      return null;
    }

    return chips.map((chip, key) =>
      onChipDelete ? (
        <Chip key={key} label={chip} marginBottom="none" onDelete={onChipDelete(chip)} />
      ) : (
        <Chip key={key} label={chip} marginBottom="none" marginTop="none" />
      ),
    );
  }
}

const InputWithForwardedRef = React.forwardRef<HTMLInputElement, InputProps>(({ className, style, ...props }, ref) => (
  <StyleableInput {...props} forwardedRef={ref} />
));

export const Input = hoistNonReactStatics(InputWithForwardedRef, StyleableInput);

Input.displayName = 'Input';
