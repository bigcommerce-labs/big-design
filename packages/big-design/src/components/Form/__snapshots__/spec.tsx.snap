// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple form render 1`] = `
.c3 {
  color: #313440;
  margin: 0 0 1rem;
  color: #5E637A;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  margin: 0 0 0.75rem;
}

.c3:last-child {
  margin-bottom: 0;
}

.c1 {
  border: none;
  margin: 0 0 1.5rem;
  padding: 0;
}

.c1:last-child {
  margin: 0;
}

.c2 {
  color: #313440;
  margin: 0 0 1rem;
  font-size: 1.25rem;
  font-weight: 600;
  line-height: 1.75rem;
  margin: 0 0 0.75rem;
}

.c2:not(:last-child) {
  margin-bottom: 0.25rem;
}

.c7 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  border-radius: 0.25rem;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 2.25rem;
  position: relative;
  width: 100%;
  border: 1px solid #D9DCE9;
}

.c7:hover:not([disabled]) {
  border: 1px solid #B4BAD1;
}

.c7[disabled] {
  background-color: #ECEEF5;
}

.c9 {
  background-color: inherit;
  border: 0;
  box-sizing: border-box;
  color: #313440;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  padding: 0;
  padding-left: 0.5rem;
  width: 100%;
  min-height: 2.125rem;
}

.c9:focus {
  outline: none;
}

.c9::-webkit-input-placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c9::-moz-placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c9:-ms-input-placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c9::placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c9:-webkit-autofill {
  background-color: #F0F3FF !important;
  -webkit-box-shadow: 0 0 0px 1000px #F0F3FF inset;
}

.c9[disabled] {
  background-color: #ECEEF5;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
}

.c4 {
  display: grid;
  grid-gap: 0.5rem 1rem;
  margin-bottom: 1rem;
}

.c4:last-child {
  margin-bottom: 0;
}

.c5 {
  color: #313440;
  margin: 0 0 1rem;
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.5rem;
  margin: 0 0 0.5rem;
  cursor: pointer;
  display: inline-block;
  margin-bottom: 0.25rem;
}

.c5::after {
  color: #5E637A;
  content: ' (optional)';
  font-weight: 400;
}

@media (min-width:720px) {
  .c0 .c6,
  .c0 .styled__StyledTextareaWrapper-c1uos0-0 {
    max-width: 26rem;
  }
}

<form
  class="c0"
>
  <fieldset
    class="c1"
  >
    <legend
      class="c2"
    >
      Primary contact
    </legend>
    <p
      class="c3"
    >
      Minim velit quis aute adipisicing adipisicing do do exercitation cupidatat enim ex voluptate consequat labore.
    </p>
    <div
      class="c4"
    >
      <div>
        <label
          class="c5"
          for="bd-input-1"
        >
          First Name
        </label>
        <p
          class="c3"
        >
          This is an example description for First Name
        </p>
        <span
          class="c6 c7"
        >
          <div
            class="c8"
          >
            <input
              class="c9"
              id="bd-input-1"
              placeholder="Placeholder text"
            />
          </div>
        </span>
      </div>
    </div>
    <div
      class="c4"
    >
      <div>
        <label
          class="c5"
          for="bd-input-2"
        >
          Middle Name
        </label>
        <p
          class="c3"
        >
          This is an example description for Last Name. Featuring a Left Icon.
        </p>
        <span
          class="c6 c7"
        >
          <div
            class="c8"
          >
            <input
              class="c9"
              id="bd-input-2"
              placeholder="Placeholder text"
            />
          </div>
        </span>
      </div>
    </div>
  </fieldset>
</form>
`;
